packages/@lwc/engine-core/src/framework/hooks.ts

```diff
@@ -164,18 +162,9 @@ export function allocateChildrenHook(vnode: VCustomElement, vm: VM) {
     //
     // In case #2, we will always get a fresh VCustomElement.
     const children = vnode.aChildren || vnode.children;
-    vm.aChildren = children;
-
-    if (process.env.NODE_ENV !== 'production') {
-        assert.isTrue(
-            hasShadow(vm) || children.length === 0,
-            `Invalid usage of ${getComponentTag(
-                vm
-            )}. Light DOM components don't support slotting yet.`
-        );
-    }
 
-    if (isTrue(vm.renderer.syntheticShadow)) {
+    vm.aChildren = children;
+    if (isTrue(vm.renderer.syntheticShadow) || !hasShadow(vm)) {
```

Comment:
Now that https://github.com/salesforce/lwc/pull/2344 is merged, it might also be useful to rename in another PR `hasShadow` to better represent that the component is rendering in the light DOM.